flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java

```diff
@@ -0,0 +1,167 @@
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.hiveAggDecimalPlus;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.ifThenElse;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.isNull;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.nullOf;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.plus;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.tryCast;
+import static org.apache.flink.table.planner.expressions.ExpressionBuilder.typeLiteral;
+import static org.apache.flink.table.types.logical.DecimalType.MAX_PRECISION;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.DECIMAL;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getPrecision;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getScale;
+
+/** built-in hive sum aggregate function. */
+public class HiveSumAggFunction extends HiveDeclarativeAggregateFunction {
+
+    private final UnresolvedReferenceExpression sum = unresolvedRef("sum");
+    private DataType argsType;
+    private DataType resultType;
+
+    @Override
+    public int operandCount() {
+        return 1;
+    }
+
+    @Override
+    public UnresolvedReferenceExpression[] aggBufferAttributes() {
+        return new UnresolvedReferenceExpression[] {sum};
+    }
+
+    @Override
+    public DataType[] getAggBufferTypes() {
+        return new DataType[] {getResultType()};
+    }
+
+    @Override
+    public DataType getResultType() {
+        return resultType;
+    }
+
+    @Override
+    public Expression[] initialValuesExpressions() {
+        return new Expression[] {/* sum = */ nullOf(getResultType())};
+    }
+
+    @Override
+    public Expression[] accumulateExpressions() {
+        Expression operand;
+        // TimestampToNumericCastRule can't cast timestamp to numeric directly, so here use
+        // UNIX_TIMESTAMP(CAST(timestamp_col AS STRING)) instead
+        if (argsType.getLogicalType().is(TIMESTAMP_WITHOUT_TIME_ZONE)) {
+            operand = castTimestampToLong(operand(0));
+        } else {
+            operand = operand(0);
+        }
+        Expression tryCastOperand = tryCast(operand, typeLiteral(getResultType()));
+        return new Expression[] {
+            /* sum = */ ifThenElse(
+                    isNull(operand(0)),
+                    sum,
+                    ifThenElse(
+                            isNull(tryCastOperand),
+                            sum,
+                            ifThenElse(
+                                    isNull(sum),
+                                    tryCastOperand,
+                                    adjustedPlus(sum, tryCastOperand))))
+        };
+    }
+
+    @Override
+    public Expression[] retractExpressions() {
```

Comment:
Why sum could not support retraction?

Reply:
We just follow hive logic that doesn't support retraction.